waiter/src/waiter/scheduler.clj

```diff
@@ -642,3 +646,182 @@
+                                          ((juxt :service-id->healthy-instances
+                                                 :service-id->unhealthy-instances))
+                                          (map #(map :id (get % service-id#)))
+                                          (reduce into #{}))]
+             (make-launch-tracker service-id# instance-counts# known-instance-ids#)))]
+     (log/info "Starting launch metrics loop on iteration" iteration#
+               "with services:" initial-service-ids#)
+     {:known-service-ids initial-service-ids#
+      :previous-iteration iteration#
+      :service-id->launch-tracker initial-service-id->launch-tracker#}))
+
+(defn start-launch-metrics-maintainer
+  "go block to collect metrics on the instance launch overhead of waiter services.
+
+  Initial state for the router is read from `router-state-query-chan`.
+  Updates to state for the router are then read from `router-state-updates-chan`."
+  [router-state-query-chan router-state-updates-chan]
+  (let [exit-chan (async/chan 1)
+        query-chan (au/latest-chan)
+        update-state-timer (metrics/waiter-timer "state" "launch-metrics-maintainer" "update-state")
+        waiter-schedule-timer (metrics/waiter-timer "launch-overhead" "schedule-time")]
+    (async/go
+      (try
+        (loop [{:keys [known-service-ids previous-iteration service-id->launch-tracker] :as current-state}
+               ;; NOTE: the iteration initial state read from router-state-query-chan is NOT guaranteed
+               ;; as <= the iteration of states later read from router-state-updates-chan.
+               ;; This is because the router-state-maintainer puts! to the channels asynchronously.
+               ;; The if-not < iteration check below to addresses this problem.
+               (build-intital-service-launch-trackers router-state-query-chan)]
+          (let [new-state
+                (async/alt!
+                  exit-chan
+                  ([message]
+                   (if (= :exit message)
+                     (do
+                       (log/warn "stopping launch-metrics-maintainer")
+                       (comment "Return nil to exit the loop"))
+                     current-state))
+
+                  router-state-updates-chan
+                  ([router-state]
+                   (if-not (< previous-iteration (:iteration router-state))
+                     current-state  ;; ignoring out-of-order state updates
+                     (timers/start-stop-time!
+                       update-state-timer
+                       (let [{:keys [iteration service-id->healthy-instances service-id->instance-counts
+                                     service-id->instance-slots service-id->my-instance->slots
+                                     service-id->unhealthy-instances]} router-state
+                             incoming-service-ids (set (keys service-id->my-instance->slots))
```

Comment:
Could we use some other field to compute `incoming-service-ids`? What is special in `service-id->my-instance->slots`?

Reply:
Nothing special, aside from the fact that one of the other state loops was doing it that way. I've removed `service-id->my-instance->slots` and now use `service-id->instance-counts` instead.